### PR TITLE
perf(sql): speed up self-join queries on symbol columns

### LINKED ISSUE #4712: ASOF self-join on symbol column uses string keys instead of int ones
### Is your feature request related to a problem?

If you run the following query on demo, it'll use a string key in the [hash table](https://github.com/questdb/questdb/blob/2246e559966f5e94e2d9bea2aec0e46c2b91adbf/core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinLightRecordCursorFactory.java#L66) instead of a more efficient int one:
```sql
select * from trades t1 asof join trades t2 on(symbol);
```

We should detect self-joins and then use `int` type for symbol columns.

This will unlock the following performance improvement (measured on clickbench dataset and 4c/8t 32GB Linux machine):
```sql
-- string key: 7.56s
-- int key: 4.26s
select h2.WatchID from hits h1 asof join hits h2 on(UserAgentMinor) limit -1;
```

### Describe the solution you'd like.

_No response_

### Describe alternatives you've considered.

_No response_

### Full Name:

Andrei Pechkurov

### Affiliation:

QuestDB

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Closes #4712

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 72 / 74 (97.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/RecordSinkFactory.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|52|53|98.11%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctTimeSeriesRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/BitSet.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByAtom.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 72 / 74 (97.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/RecordSinkFactory.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|52|53|98.11%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctTimeSeriesRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/BitSet.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByAtom.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java|1|1|100.00%|

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java [hunk 1]
```diff
@@ -1245,6 +1245,252 @@ public void testNestedLT_keySymbol_moreColumns() throws Exception {
...
+                    "INSERT INTO x VALUES " +
+                            "('1', '2', '2000-01-01T00:00:00.000000Z')," +
+                            "('3', '4', '2000-01-01T00:00:00.000000Z')," +
+                            "('1', '1', '2000-01-01T00:00:00.000000Z')," +
+                            "('2', '2', '2000-01-01T00:00:00.000000Z')," +
+                            "('4', '3', '2000-01-01T00:00:00.000000Z')"
+            );
+
+            // ASOF JOIN
+            String query = "SELECT * FROM (select sym1 s, ts from x) x1 " +
```
Comment (ideoma) [inline core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:1416]: That's an interesting case, but the dates are all the same, it's hard to say what records are joined.
Can you make records distinguishable by making the timestamp unique or adding another column?

**Review by ideoma — APPROVED** [review]

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:1474]: Changed the dates to be different in 28c3d72c0edf177e3ac9ca5da67dbad547951874

File: core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java [hunk 2]
```diff
@@ -1245,6 +1245,252 @@ public void testNestedLT_keySymbol_moreColumns() throws Exception {
...
+            // SPLICE JOIN
+            query = "SELECT * FROM (select sym1 s, ts from x) x1 " +
+                    "SPLICE JOIN (select sym2 s, ts from x) x2 ON(s)";
+            expected = "s\tts\ts1\tts1\n" +
+                    "1\t2000-01-01T00:00:00.000000Z\t\t\n" +
+                    "3\t2000-01-01T00:00:00.000000Z\t\t\n" +
+                    "1\t2000-01-01T00:00:00.000000Z\t1\t2000-01-01T00:00:00.000000Z\n" +
+                    "2\t2000-01-01T00:00:00.000000Z\t2\t2000-01-01T00:00:00.000000Z\n" +
+                    "4\t2000-01-01T00:00:00.000000Z\t\t\n";
+            assertQueryNoLeakCheck(expected, query, null, false, false);
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/test/java/io/questdb/test/griffin/AsOfJoinTest.java:1504]: Added a similar test, but with different tables in 28c3d72c0edf177e3ac9ca5da67dbad547951874

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 72 / 74 (97.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/RecordSinkFactory.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|52|53|98.11%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctTimeSeriesRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/BitSet.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByAtom.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java|1|1|100.00%|

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 72 / 74 (97.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/RecordSinkFactory.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|52|53|98.11%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctTimeSeriesRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByFillNoneRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/BitSet.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/DistinctRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByAtom.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/AbstractSampleByFillRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java|1|1|100.00%|

**Pull request merged**